src/title.js

```diff
@@ -0,0 +1,441 @@
+		// Title always has to be the first element in the root.
+		for ( const rootChild of Array.from( modelRoot.getChildren() ) ) {
+			// If the first element is not a title we need to fix it.
+			if ( rootChild.index === 0 && !rootChild.is( 'title' ) ) {
+				const title = this._getTitleElement();
+
+				// Change first element to the title if it can be a title.
+				if ( Title.titleLikeElements.has( rootChild.name ) ) {
+					const position = model.createPositionBefore( rootChild );
+					const title = writer.createElement( 'title' );
+
+					writer.insert( title, position );
+					writer.append( rootChild, title );
+					writer.rename( rootChild, 'title-content' );
+
+					// After changing element to the title it has to be filtered out from disallowed attributes.
+					model.schema.removeDisallowedAttributes( [ rootChild ], writer );
+
+				// If the first element cannot be a title but title is already in the root
+				// than move the first element after a title.
+				} else if ( title ) {
+					const positionAfterTitle = writer.createPositionAt( title, 'after' );
+
+					// To preserve correct order when more than one element is moved in the one post-fixer call.
+					const targetPosition = positionAfterTitle.getShiftedBy( numberOfMovedElements );
+
+					writer.move( writer.createRangeOn( rootChild ), targetPosition );
+					numberOfMovedElements++;
+
+				// If there is no title or any element that could be a title then create an empty title.
+				} else {
+					writer.insertElement( 'title', modelRoot );
+					writer.insertElement( 'title-content', modelRoot.getChild( 0 ) );
+				}
+
+				hasChanged = true;
+
+			// If there is a title somewhere in the content.
+			} else if ( rootChild.index > 0 && rootChild.is( 'title' ) ) {
+				// Rename it to a paragraph if it has a content.
+				if ( model.hasContent( rootChild ) ) {
+					writer.rename( rootChild, 'paragraph' );
+					writer.unwrap( rootChild.getChild( 0 ) );
+				// Or remove if it's empty (it's created as a result of splitting parents by insertContent method).
+				} else {
+					writer.remove( rootChild );
+				}
+
+				hasChanged = true;
+			}
+		}
+
+		// Attach `Body` placeholder when there is no element after a title.
+		if ( modelRoot.childCount < 2 ) {
+			this._bodyPlaceholder = writer.createElement( 'paragraph' );
+			writer.insert( this._bodyPlaceholder, modelRoot, 1 );
+			hasChanged = true;
+		}
+
+		return hasChanged;
+	}
+
+	/**
+	 * Prevents editor from adding extra paragraphs after pasting to the title element
+	 * or pressing an enter in the title element.
+	 *
+	 * @private
+	 * @param {module:engine/model/writer~Writer} writer
+	 * @returns {Boolean}
+	 */
+	_preventExtraParagraphing( writer ) {
+		const root = this.editor.model.document.getRoot();
+		const placeholder = this._bodyPlaceholder;
+
+		const shouldDeleteLastParagraph = (
+			placeholder && placeholder.is( 'paragraph' ) && placeholder.childCount === 0 &&
```

Comment:
please refactor this case as this is tooo long :)

Maybe early return when no placeholder (first check).

ps.: A more accurate name would be : `shouldRemovePlaceholder` ;)